tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,767 @@
+	HeadChunkFileHeaderSize = SegmentHeaderSize
+	// MaxHeadChunkFileSize is the max size of a head chunk file.
+	MaxHeadChunkFileSize = 512 * 1024 * 1024 // 512 MiB.
+	// CRCSize is the size of crc32 sum on disk.
+	CRCSize = 4
+	// MaxHeadChunkMetaSize is the max size of an mmapped chunks minus the chunks data.
+	// Max because the uvarint size can be smaller.
+	MaxHeadChunkMetaSize = SeriesRefSize + 2*MintMaxtSize + ChunksFormatVersionSize + MaxChunkLengthFieldSize + CRCSize
+)
+
+// corruptionErr is an error that's returned when corruption is encountered.
+type corruptionErr struct {
+	Dir       string
+	FileIndex int
+	Err       error
+}
+
+func (e *corruptionErr) Error() string {
+	return errors.Wrapf(e.Err, "corruption in head chunk file %s", segmentFile(e.Dir, e.FileIndex)).Error()
+}
+
+// ChunkDiskMapper is for writing the Head block chunks to the disk
+// and access chunks via mmapped file.
+type ChunkDiskMapper struct {
+	/// Writer.
+	dir *os.File
+
+	curFile         *os.File // File being written to.
+	curFileSequence int      // Index of current open file being appended to.
+	curFileMint     int64    // Used to check for a chunk crossing the max file time range.
+	curFileMaxt     int64    // Used for the size retention.
+	curFileNumBytes int64    // Bytes written in current open file.
+	maxFileTime     int64    // Max time range (curFileMaxt-curFileMint) for a file.
+
+	byteBuf      [MaxHeadChunkMetaSize]byte // Buffer used to write the header of the chunk.
+	chkWriter    *bufio.Writer              // Writer for the current open file.
+	crc32        hash.Hash
+	writePathMtx sync.RWMutex
+
+	/// Reader.
+	// The int key in the map is the file number on the disk.
+	mmappedChunkFiles map[int]*mmappedChunkFile // Contains the m-mapped files for each chunk file mapped with its index.
+	closers           map[int]io.Closer         // Closers for resources behind the byte slices.
+	readPathMtx       sync.RWMutex              // Mutex used to protect the above 2 maps.
+	pool              chunkenc.Pool             // This is used when fetching a chunk from the disk to allocate a chunk.
+
+	// Writer and Reader.
+	// We flush chunks to disk in batches. Hence, we store them in this buffer
+	// from which chunks are served till they are flushed and are ready for m-mapping.
+	chunkBuffer *chunkBuffer
+
+	// The total size of bytes in the closed files.
+	// Needed to calculate the total size of all segments on disk.
+	size int64
+
+	// If 'true', it indicated that the maxt of all the on-disk files were set
+	// after iterating through all the chunks in those files.
+	fileMaxtSet bool
+
+	closed bool
+}
+
+type mmappedChunkFile struct {
+	byteSlice ByteSlice
+	maxt      int64
+}
+
+// NewChunkDiskMapper returns a new writer against the given directory
+// using the default head chunk file duration.
+// NOTE: 'IterateAllChunks' method needs to be called at least once after creating ChunkDiskMapper
```

Comment:
This is weird - just run IterateAll Chunks in constructor? ;p

Reply:
Yeah, that would have been good, but, `Head` would not like it :P. The initialization and replay are separated in `Head` (`NewHead` and `Init`), so we need to have it separately here.